File: app/src/main/java/com/anthonysottile/kenken/settings/SettingsProvider.java
package com.anthonysottile.kenken.settings;

import android.content.SharedPreferences;

import com.anthonysottile.kenken.ui.UIConstants;

import java.util.ArrayList;
import java.util.EventListener;
import java.util.EventObject;
import java.util.List;

public final class SettingsProvider {

    private static final String GameSize = "GameSize";

    private static SharedPreferences preferences = null;

    private static int gameSize = UIConstants.MinGameSize;

    // TODO: implement dirty cache methodology

    public interface GameSizeChangedListener extends EventListener {
        void onGameSizeChanged(EventObject event);
    }

    private static final List<GameSizeChangedListener> gameSizeChangedListeners =
            new ArrayList<>();

    public static void AddGameSizeChangedListener(GameSizeChangedListener listener) {
        SettingsProvider.gameSizeChangedListeners.add(listener);
    }

    private static void triggerGameSizeChanged() {
        EventObject event = new EventObject(new Object());

        for (GameSizeChangedListener listener : SettingsProvider.gameSizeChangedListeners) {
            listener.onGameSizeChanged(event);
        }
    }

    /**
     * Returns the setting for the game size.
     *
     * @return The setting for the game size.
     */
    public static int GetGameSize() {
        return SettingsProvider.gameSize;
    }

    /**
     * Sets the game size setting.  Triggers game size changed if it changes.
     *
     * @param gameSize The size of the game to set.
     */
    public static void SetGameSize(int gameSize) {
        if (gameSize != SettingsProvider.gameSize) {
            SettingsProvider.gameSize = gameSize;
            SharedPreferences.Editor editor = SettingsProvider.preferences.edit();
            editor.putInt(SettingsProvider.GameSize, SettingsProvider.gameSize);
            editor.apply();

            SettingsProvider.triggerGameSizeChanged();
        }
    }

    /**
     * Initialization method to give the manager a reference to preferences.
     *
     * @param preferences A reference to the application preferences.
     */
    public static void Initialize(SharedPreferences preferences) {
        SettingsProvider.preferences = preferences;
        SettingsProvider.gameSize =
                SettingsProvider.preferences.getInt(
                        SettingsProvider.GameSize,
                        UIConstants.MinGameSize
                );
    }

    private SettingsProvider() {
    }
}

File: app/src/main/java/com/anthonysottile/kenken/settings/SettingsProvider.kt
package com.anthonysottile.kenken.settings

import android.content.SharedPreferences
import com.anthonysottile.kenken.ui.UIConstants
import java.util.*

object SettingsProvider {
    private const val GameSize = "GameSize"
    private const val HardMode = "HardMode"

    private lateinit var preferences: SharedPreferences

    var gameSize = UIConstants.MinGameSize
        set (gameSize) {
            if (field != gameSize) {
                field = gameSize

                val editor = this.preferences.edit()
                editor.putInt(this.GameSize, this.gameSize)
                editor.apply()

                SettingsProvider.triggerGameSizeChanged()
            }
        }

    var hardMode = false
        set (hardMode) {
            if (field != hardMode) {
                field = hardMode

                val editor = this.preferences.edit()
                editor.putBoolean(this.HardMode, this.hardMode)
                editor.apply()

                SettingsProvider.triggerHardModeChanged()
            }
        }

    private val gameSizeChangedListeners = ArrayList<() -> Unit>()

    // TODO: implement dirty cache methodology

    fun addGameSizeChangedListener(listener: () -> Unit) {
        SettingsProvider.gameSizeChangedListeners.add(listener)
    }

    private fun triggerGameSizeChanged() {
        for (listener in SettingsProvider.gameSizeChangedListeners) {
            listener()
        }
    }

    private val hardModeChangedListeners = ArrayList<() -> Unit>()

    fun addHardModeChangedListener(listener: () -> Unit) {
        SettingsProvider.hardModeChangedListeners.add(listener)
    }

    private fun triggerHardModeChanged() {
        for (listener in SettingsProvider.hardModeChangedListeners) {
            listener()
        }
    }

    /**
     * Initialization method to give the manager a reference to preferences.
     *
     * @param preferences A reference to the application preferences.
     */
    fun initialize(preferences: SharedPreferences) {
        this.preferences = preferences
        this.gameSize = SettingsProvider.preferences.getInt(
                SettingsProvider.GameSize,
                UIConstants.MinGameSize
        )
        this.hardMode = SettingsProvider.preferences.getBoolean(
                SettingsProvider.HardMode,
                false
        )
    }
}
